pkg/chunk/local/boltdb_index_client.go

```diff
@@ -45,18 +52,71 @@ func NewBoltDBIndexClient(cfg BoltDBConfig) (chunk.IndexClient, error) {
 		return nil, err
 	}
 
-	return &boltIndexClient{
-		cfg: cfg,
-		dbs: map[string]*bbolt.DB{},
-	}, nil
+	indexClient := &boltIndexClient{
+		cfg:  cfg,
+		dbs:  map[string]*bbolt.DB{},
+		done: make(chan struct{}),
+	}
+
+	indexClient.wait.Add(1)
+	go indexClient.loop()
+	return indexClient, nil
+}
+
+func (b *boltIndexClient) loop() {
+	defer b.wait.Done()
+
+	ticker := time.NewTicker(dbReloadPeriod)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			b.reload()
+		case <-b.done:
+			return
+		}
+	}
+}
+
+func (b *boltIndexClient) reload() {
```

Comment:
> We open any unopened dbs and add them to current set.
Do we want to do this? Not all the dbs needs to be opened I feel since queriers would be limited to only a few dbs.

We still would have to lock `dbsMtx`, for finding deleted dbs we can have read lock, and if there are any deleted dbs then we can have exclusive lock. What do you think?

Reply:
Yes, just that we don't block queries unless we need to.